File: 103-binary-tree-zigzag-level-order-traversal/103-binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
	public static List<List<Integer>> zigzagLevelOrder(TreeNode root) {
		Queue<TreeNode> q = new LinkedList<>();

		List<List<Integer>> return_list = new LinkedList<List<Integer>>();

		if (root == null) {
			return return_list;
		}

		q.add(root);
		
		boolean odd = true;
		
		while(!q.isEmpty()) {
			int size = q.size();
			List<Integer> temp_list = new LinkedList<>();
			
			for(int i = 0; i < size; i++) {
				TreeNode currentNode = q.poll();
				
				if(odd)
					temp_list.add(currentNode.val);
				else
					temp_list.add(0, currentNode.val);
				
				if (currentNode.left != null)
					q.add(currentNode.left);
				if (currentNode.right != null)
					q.add(currentNode.right);
	
			}
		
			return_list.add(temp_list);
			if(odd)
				odd = false;
			else
				odd =true;
			
		}
		return return_list;
	}

}

File: 103-binary-tree-zigzag-level-order-traversal/103-binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        list = []

        if not root:
            return list

        queue = [root]

        even = False

        while queue:
            even = not even
            temp = []
            size = len(queue)

            for i in range(size):
                curr = queue.pop(0)

                if even:
                    temp.append(curr.val)
                else:
                    temp.insert(0, curr.val)

                if curr.left:

                    queue.append(curr.left)
                if curr.right:
                    queue.append(curr.right)

            list.append(temp)

        return list
